libraries/core_libs/network/src/tarcap/packets_handlers/get_pbft_sync_packet_handler.cpp

```diff
@@ -32,38 +32,58 @@ void GetPbftSyncPacketHandler::process(const PacketData &packet_data,
-    blocks_to_transfer = std::min(network_sync_level_size_, (my_chain_size - (height_to_sync - 1)));
+    auto total_sync_blocks_size = my_chain_size - height_to_sync + 1;
+    if (total_sync_blocks_size <= network_sync_level_size_) {
+      blocks_to_transfer = total_sync_blocks_size;
+      will_be_synced = true;
+    } else {
+      blocks_to_transfer = network_sync_level_size_;
+    }
+    LOG(log_tr_) << "Will send " << blocks_to_transfer << " PBFT blocks to " << packet_data.from_node_id_;
+  } else {
+    LOG(log_wr_) << "Request syncing period is ahead of own PBFT chain. Peer " << packet_data.from_node_id_
+                 << " want to sync PBFT blocks period from " << height_to_sync << ". But own PBFT chain size is "
+                 << my_chain_size;
+    will_be_synced = true;
   }
-  LOG(log_tr_) << "Will send " << blocks_to_transfer << " PBFT blocks to " << packet_data.from_node_id_;
 
-  sendPbftBlocks(packet_data.from_node_id_, height_to_sync, blocks_to_transfer);
+  sendPbftBlocks(packet_data.from_node_id_, height_to_sync, blocks_to_transfer, will_be_synced);
 }
 
 // api for pbft syncing
 void GetPbftSyncPacketHandler::sendPbftBlocks(dev::p2p::NodeID const &peer_id, size_t height_to_sync,
-                                              size_t blocks_to_transfer) {
+                                              size_t blocks_to_transfer, bool will_be_synced) {
   LOG(log_tr_) << "sendPbftBlocks: peer want to sync from pbft chain height " << height_to_sync
                << ", will send at most " << blocks_to_transfer << " pbft blocks to " << peer_id;
-  uint64_t current_period = height_to_sync;
+  dev::RLPStream s;
+  auto pbft_chain_synced = false;
   if (blocks_to_transfer == 0) {
-    LOG(log_dg_) << "Sending empty PbftSyncPacket to " << peer_id;
-    sealAndSend(peer_id, SubprotocolPacketType::PbftSyncPacket, dev::RLPStream(0));
+    pbft_chain_synced = true;
+    s << pbft_chain_synced;
+    sealAndSend(peer_id, SubprotocolPacketType::PbftSyncPacket, std::move(s));
     return;
   }
 
+  uint64_t current_period = height_to_sync;
   while (current_period < height_to_sync + blocks_to_transfer) {
     bool last_block = (current_period == height_to_sync + blocks_to_transfer - 1);
     auto data = db_->getPeriodDataRaw(current_period);
 
     if (data.size() == 0) {
-      sealAndSend(peer_id, SubprotocolPacketType::PbftSyncPacket, dev::RLPStream(0));
+      pbft_chain_synced = true;
+      s << pbft_chain_synced;
+      sealAndSend(peer_id, SubprotocolPacketType::PbftSyncPacket, std::move(s));
       LOG(log_er_) << "Missing pbft block in db, send no pbft blocks to " << peer_id;
       return;
     }
 
-    dev::RLPStream s;
-    s.appendList(2);
+    if (will_be_synced && last_block) {
+      pbft_chain_synced = true;
+    }
+    s.appendList(3);
+    s << pbft_chain_synced;
     s << last_block;
     s.appendRaw(data);
     LOG(log_dg_) << "Sending PbftSyncPacket period " << current_period << " to " << peer_id;
```

Comment:
line under this is huge BUG, you can not use `sealAndSend(peer_id, SubprotocolPacketType::PbftSyncPacket, std::move(s));` because `s` is defined outside of this loop and you can not move it several times

Reply:
fixed, thanks